### PR TITLE
Enable mixing intra and inter blocks in inter frames

### PR DESCRIPTION
**Pull request by fbossen**

### CODE REVIEW

File: src/context.rs [hunk 1]
```diff
@@ -1091,29 +1091,47 @@ impl BlockOffset {
...
+      y: y as usize
+    }
+  }
 }
 
 #[derive(Copy, Clone)]
 pub struct Block {
   pub mode: PredictionMode,
-  pub bsize: BlockSize,
+  //pub bsize: BlockSize,
```
Comment (tdaede) [inline src/context.rs]: drop these commented out bits

File: src/context.rs [hunk 2]
```diff
@@ -1091,29 +1091,47 @@ impl BlockOffset {
...
+  pub n8_h: usize, /* block height in the unit of mode_info */
+  pub is_sec_rect: bool,
 }
 
 impl Block {
   pub fn default() -> Block {
     Block {
       mode: PredictionMode::DC_PRED,
-      bsize: BlockSize::BLOCK_64X64,
+      //bsize: BlockSize::BLOCK_64X64,
```
Comment (tdaede) [inline src/context.rs]: same

**Comment by ycho** [comment]
If possible, please don't squash into one commit.

File: src/context.rs [hunk 3]
```diff
@@ -1091,29 +1091,47 @@ impl BlockOffset {
...
+      //sb_type: BlockSize::BLOCK_64X64,
       partition: PartitionType::PARTITION_NONE,
       skip: false,
       ref_frames: [INTRA_FRAME; 2],
       neighbors_ref_counts: [0; TOTAL_REFS_PER_FRAME],
-      cdef_index: 0
+      cdef_index: 0,
+      n8_w: BlockSize::MI_SIZE_WIDE[BLOCK_64X64 as usize],
```
Comment (tdaede) [inline src/context.rs]: I think you can just to BLOCK_64X64.width_mi() here

File: src/context.rs [hunk 4]
```diff
@@ -1748,6 +1776,345 @@ impl ContextWriter {
     symbol_with_update!(self, w, enable as u32, &mut self.fc.filter_intra_cdfs[block_size as usize]);
   }
 
+  fn get_mvref_ref_frames(&mut self, ref_frame: usize) -> ([usize; 2], usize) {
+    let ref_frame_map: [[usize; 2]; TOTAL_COMP_REFS] = [
+      [ LAST_FRAME,  BWDREF_FRAME  ], [ LAST2_FRAME,  BWDREF_FRAME  ],
```
Comment (tdaede) [inline src/context.rs]: This is fine for now but I'd rather drop all the *_FRAME constants at some point, maybe just with a find/replace here.

File: src/context.rs [hunk 5]
```diff
@@ -1097,23 +1097,29 @@ impl BlockOffset {
...
       bsize: BlockSize::BLOCK_64X64,
+      sb_type: BlockSize::BLOCK_64X64,
       partition: PartitionType::PARTITION_NONE,
       skip: false,
       ref_frames: [INTRA_FRAME; 2],
       neighbors_ref_counts: [0; TOTAL_REFS_PER_FRAME],
-      cdef_index: 0
+      cdef_index: 0,
+      n8_w: BlockSize::MI_SIZE_WIDE[BLOCK_64X64 as usize],
+      n8_h: BlockSize::MI_SIZE_HIGH[BLOCK_64X64 as usize],
```
Comment (ycho) [inline src/context.rs]: Copying the comment I've left at atomnuker's branch.
"You probably know n8_w and n8_h means n4_w and n4_h?"
I've fixed most of these outdated names in aomenc, but missed many in aomdec."

File: src/context.rs [hunk 6]
```diff
@@ -1097,23 +1097,29 @@ impl BlockOffset {
...
+  pub n8_w: usize, /* block width in the unit of mode_info */
+  pub n8_h: usize, /* block height in the unit of mode_info */
 }
 
 impl Block {
   pub fn default() -> Block {
     Block {
       mode: PredictionMode::DC_PRED,
       bsize: BlockSize::BLOCK_64X64,
+      sb_type: BlockSize::BLOCK_64X64,
```
Comment (ycho) [inline src/context.rs]: Copying the comment I've left at atomnuker's branch.
"The existing 'bsize' is exactly the same to newly added sb_type."

File: src/context.rs (hunk 5 above)
Comment (fbossen) [reply src/context.rs]: Yes, these names should be changed. Even n4_w and n4_h may be confusing. Can we do this refactoring in a separate PR?

File: src/context.rs (hunk 6 above)
Comment (fbossen) [reply src/context.rs]: Yes. Both have been removed at this time since they are not used.

File: src/context.rs (hunk 5 above)
Comment (ycho) [reply src/context.rs]: Sure- let's correct them in separate patch. There are several more defined locally.

File: src/context.rs (hunk 4 above)
Comment (tdaede) [reply src/context.rs]: I agree

**Pull request merged**